### ISSUE TITLE
Use table-level display options to allow users to re-order columns

### ISSUE THREAD

**Issue by seancolsen** [centerofci/mathesar#1584]
## Current behavior

- Columns are displayed within the Table Page and within the embedded Table Widget on the Record Page, and probably other places too.
- Columns are always ordered by their immutable Postgres attnum value (i.e. the order in which they were created). The user has no way to display newly-added columns prior to previously added columns.
- Some users may wish to modify the order in which columns are displayed.

## Desired behavior

- Within the Table Page, users can drag columns to reorder them.
- Within the table widgets (displayed on the Record Page), users can _not_ modify the column ordering (at least for the time being). The rationale here is that, because we intend to hide the linking PK column, it's possible the a reordering operation may be ambiguous without all columns present.

## UX

- The UX should be identical to Google Sheets -- you need to select the column before dragging it. When no cells are selected, initiating a drag event via the column header will begin selecting cells in multiple columns, allowing the user to select all cells in adjacent columns via dragging. When all cells within a particular column are selected, then that column's header becomes a target to begin a drag event which re-orders the column. The cursor will provide visual feedback as to the differences in behavior before dragging. This is a little hard to explain. Just try it on Google Sheets and you'll see how it works.

## Implementation

- The user-defined order is persisted via a table-level `display_options` property, `colum_order`, which is an array of `attnum` values.
- If the user has not customized the column order, then `colum_order` will be `null`.
- The order of columns (or cells) within API responses is undefined. Columns and cells can be returned in whatever order is convenient. The front end is responsible for sorting columns and cells into their appropriate order.
- Any time we display the columns within a table, we first display all the columns as listed in `column_order` (keeping in-mind that `column_order` may contain values for deleted columns which need to be filtered out), then we display any remaining columns, as sorted by their `attnum` values.
- When a user re-orders columns, the front end submits one PATCH request to update `colum_order`, providing the full array containing all `attnum` values.
- The above approach allows us to keep the columns metadata in Django thin and decoupled from Postgres. The computed order is deterministic and able to withstand DDL changes made outside of Mathesar without the need for reflection.
- The `colum_order` display option does not affect the selection of a column for the Record Summary. If a user wishes to change the column automatically selected for the Record Summary, they can do so by providing a custom summary template (not by re-ordering the columns).
- The `column_order` display option is not exposed directly via any UI, beyond the order in which the columns display.
- The `column_order` customization should apply identically for all users.

## Task scheduling

- This feature requires back end work and front end work. That work can be taken up in either order. If you want to implement the front end first, then begin by allowing the user to reorder columns without that order persisting via the API.

**Comment by AlekLefebvre** [comment]
I'd like to help with this issue.

**Comment by seancolsen** [comment]
This is going to be a heavier lift than the other things you've worked on so far, especially since it will require both front end changes _and_ back end changes (in Python). If you're up to the challenge, we'd love to have your help! Feel free to begin work and submit a PR when you're ready. I would strongly recommend submitting a back-end only PR first so that we can review/merge it before you spend time on the front end portion.

**Comment by AlekLefebvre** [comment]
@seancolsen Yes, no problem working in Python. I'll prepare a PR for the back end first.

**Comment by seancolsen** [comment]
We had to revert the changes in #2213 temporarily to make a release without some regressions, so I'm re-opening this issue and assigning it to @AlekLefebvre who [expressed interest](https://github.com/centerofci/mathesar/pull/2213#issuecomment-1480352808) in continuing work.

Continued progress is currently blocked as we are waiting for the 0.1.1 release to be merged into master and develop. @AlekLefebvre I'll comment here when it's ready.

**Comment by seancolsen** [comment]
@AlekLefebvre We've completed the 0.1.1 release so now you can resume work. Here's how:

1. Create a new branch based on the latest version of `develop`.

1. Run 

    ```
    git revert -m 1 ef067446a9c21191140d50c88eb67638167248d4
    ```

    That will make a new commit on your branch with all the changes you've made thus far on this feature.

1. Then push additional commits to address the regression described in #2717, #2718, and #2722.

1. Then open a new PR which targets this issue.
